{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Setting up the code, importing all necessary packages and data\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from Bio import Entrez\n",
    "import os\n",
    "import time\n",
    "Entrez.email = \"spatuzzi@stud.uni-heidelberg.de\"\n",
    "\n",
    "#Modify this string based on your working directory in which your file is stored\n",
    "\n",
    "WorkingDirectory = '/Users/darthvader/Desktop/Heidelberg/1_MoBi_Master/Praktikum_Bromham/Data'\n",
    "os.chdir(WorkingDirectory)\n",
    "Phy_genes = pd.read_csv(WorkingDirectory+'/Phy_genes.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Data manipulation, establish global objects that will be needed later\n",
    "\n",
    "#Dictionary to pair each gene with its \"product\" name\n",
    "Product_dict = {\"12S\":[\"12S ribosomal RNA\"], \"16S\":[\"16S ribosomal RNA\"],\"BDNF\":[\"brain-derived neurotrophic factor\"], \"CXCR4\":[\"chemokine receptor 4\"],\n",
    "            \"cytb\":[\"cytochrome b\"], \"H3A\":[\"histone+H3\", \"histone H3a\"], \"NCX1\":[\"sodium/calcium exchanger 1\"], \"ND1\":[\"NADH dehydrogenase subunit 1\",\n",
    "            \"ND1\"], \"ND2\":[\"NADH dehydrogenase subunit 2\"], \"POMC\":[\"proopiomelanocortin\"], \"RAG1\":[\"recombination activating protein 1\"], \"RHOD\":[\"rhodopsin\"],\n",
    "             \"SIA\":[\"seventh in absentia\"], \"SLC8A3\":[\"solute carrier family 8 member 3\"], \"TYR\":[\"tyrosinase\"]}\n",
    "Product_dict_key = Product_dict.keys()\n",
    "\n",
    "#Edit the Phy_genes Dataframe\n",
    "#We change to row names to the scientific names in order to use the .loc method, but do not get rid of the \"Scientific name\" \n",
    "#column because we will revert to a numbered index when we have multiple sequences for some gene-organism match\n",
    "Phy_genes.columns = ['Scientific name', '12S', '16S', 'BDNF', 'CXCR4', 'cytb', 'H3A', 'NCX1',\n",
    "       'ND1', 'ND2', 'POMC', 'RAG1', 'RHOD', 'SIA', 'SLC8A3', 'TYR']\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Replace empty values with \"None\"\n",
    "\n",
    "Phy_genes.index = Phy_genes.iloc[:,0].values\n",
    "\n",
    "for name in Phy_genes.index:\n",
    "    for gene in Phy_genes.columns:\n",
    "        if(Phy_genes.loc[name, gene].count(\"-\") == len(Phy_genes.loc[name, gene])):\n",
    "                Phy_genes.loc[name,gene] = None\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create new Dataframes to fill out with new results\n",
    "#Phy_genes_update: A copy of Phy_genes with the addition of new data\n",
    "#new_seq_df: Datframe containing ONLY the new data\n",
    "\n",
    "Phy_genes_update = Phy_genes.copy()\n",
    "new_seq_df = pd.DataFrame( np.empty((len(Phy_genes.index), len(Phy_genes.columns)),dtype=pd.Timestamp), index = Phy_genes.index, columns = Phy_genes.columns)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Code for iteration\n",
    "\n",
    "#Iterate over ever species\n",
    "for name in Phy_genes.index:\n",
    "    \n",
    "    #Iterate over every Gene\n",
    "    for gene in Product_dict_key:\n",
    "        \n",
    "        #Check if there is already a sequence in this position by removing all \"gaps\"\n",
    "        #If it's only gaps, continue\n",
    "        \n",
    "        if(Phy_genes.loc[name, gene] == None):\n",
    "            \n",
    "            #To keep better track in the console, announce every time gene and organism\n",
    "            print(\"Name:\" + name + \", Gene:\" + gene )\n",
    "\n",
    "            #Concatenate the term for the Entrez function and preoare empty list for the results\n",
    "            \n",
    "            term = name + \"[Orgn] AND \" + Product_dict[gene][0] + \"[Prd]\"\n",
    "            ID_list = []\n",
    "            \n",
    "            #This iteration can easily be interrupted by a temporary connection issue at this step\n",
    "            #The following chunk of code ensures that the code is run multiple times (up to 100 times) to attempt to download entrez IDs \n",
    "\n",
    "            max_tries = 100\n",
    "            for i in range(max_tries):\n",
    "                try:\n",
    "                    time.sleep(0.1) \n",
    "                    \n",
    "                    #Find IDs that match the \"product\" string\n",
    "                    \n",
    "                    handle = Entrez.esearch(db=\"nucleotide\", term = term, idtype=\"acc\")\n",
    "                    record = Entrez.read(handle)\n",
    "                    ID_list = record[\"IdList\"]\n",
    "                    print(ID_list)\n",
    "\n",
    "                    break\n",
    "                except Exception:\n",
    "                    print(\"retry\")\n",
    "                    continue\n",
    "\n",
    "            if(len(ID_list) > 0):\n",
    "\n",
    "                #Take sequence for each ID with Entrez.efetch function\n",
    "                #Edit the downloaded string object to only contain the coding sequence\n",
    "\n",
    "                seq_list = []\n",
    "                Seq_clean = \"\"\n",
    "                for ID in ID_list: \n",
    "\n",
    "                    handle = Entrez.efetch(db=\"nucleotide\", id=ID, rettype=\"gb\", retmode=\"text\")\n",
    "                    Gene_entry = handle.read()\n",
    "                    Seq_raw = Gene_entry[Gene_entry.find(\"translation\") + 13:Gene_entry.find(\"ORIGIN\")-2]\n",
    "                    Seq_raw = \"\".join(Seq_raw.split(\"\\n\"))\n",
    "                    Seq_clean = Seq_raw\n",
    "                    for remove in \" 1234567890/\":\n",
    "                        Seq_clean = Seq_clean.replace(remove, \"\").upper()\n",
    "\n",
    "                    seq_list.append(Seq_clean)\n",
    "                    \n",
    "                \n",
    "                #\"Empty\" result lists still contain a warning string, in which case they must not be appended to the result\n",
    "                if(len(seq_list) > 0 and seq_list != ['EDIDPARAMETERISEMPTY.']):\n",
    "                    Phy_genes_update.loc[name, gene] = seq_list\n",
    "                    new_seq_df.loc[name, gene] = seq_list\n",
    "                    print(\" Replaced:\" + gene + \" for \" + name)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write files for MAFFT\n",
    "import os \n",
    "\n",
    "#Create new directory\n",
    "os.system(\"mkdir New_MAlign_AA\")\n",
    "os.chdir(WorkingDirectory + \"/New_MAlign_AA\")\n",
    "\n",
    "new_seq_df.loc[:,\"Scientific name\"] = new_seq_df.index\n",
    "\n",
    "#Iterate over every gene, for each write a file with each new sequence for each organism. Organisms that have multiple sequences for the same gene \n",
    "# will have an  additional cypher (1,2,3...etc.) to differentiate them.\n",
    "for gene in Phy_genes.columns[1:17]:\n",
    "    \n",
    "    with open('New_alignment_sequences_AA_'+gene, 'w') as f:\n",
    "        seq_str = \"\"\n",
    "        for name in Phy_genes.index:\n",
    "            if(new_seq_df.loc[name, gene] != None):\n",
    "                \n",
    "                line_list = [\"> \" + name + \" \"+ str(i)+ \"\\n\" + new_seq_df.loc[name, gene][i] for i in range(len(new_seq_df.loc[name, gene]))]\n",
    "                \n",
    "                line =(\"\\n\".join(line_list))\n",
    "                f.write(line + \"\\n\")\n",
    "                              \n",
    "    \n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
